### PR TITLE
Added new max-nesting-depth rule

### LINKED ISSUE #4276: add rule : Control flow statements "if", "for", "while", "switch", "try", ... should not be nested too deeply
### Rule Suggestion

**Is your rule for a general problem or is it specific to your development style?**
General

**What does your suggested rule do?**
it helps readability of the code

**List several examples where your rule could be used**
Aerospace, Avionic, ... This rules is required for european space industry

**Additional context**
https://rules.sonarsource.com/c/RSPEC-134
https://rules.sonarsource.com/java/RSPEC-134

### PR DESCRIPTION
**Pull request by DamienGarrido**
#### PR checklist

- [x] Addresses an existing issue: fixes #4276
- [x] New feature, bugfix, or enhancement
  - [x] Includes tests
- [x] Documentation update

#### Overview of change:
Add a new `max-nesting-depth` rule (equivalent to the ESLint `max-depth` [rule](https://eslint.org/docs/rules/max-depth "ESLint max-depth rule")):
Control flow statements "if", "for", "while", "switch", "try", ... should not be nested too deeply.

#### Is there anything you'd like reviewers to focus on?

- Verify that all test cases are addressed.
- Verify the documentation.

#### CHANGELOG.md entry:
[new-rule] `max-nesting-depth`

### CODE REVIEW

File: tslint.json [hunk 1]
```diff
@@ -1,73 +1,64 @@
 {
```
Comment (JoshuaKGoldberg) [inline tslint.json]: 😕 did Prettier automatically apply these changes for you? If so, that's ok, but if not, could you please revert the unrelated lines?

File: src/rules/maxNestingDepthRule.ts [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+    }
+
+    public isEnabled(): boolean {
+        // Disable the rule if the option is provided but non-numeric or less than the minimum.
+        const isThresholdValid =
+            typeof this.threshold === "number" && this.threshold >= Rule.MINIMUM_THRESHOLD;
+        return super.isEnabled() && isThresholdValid;
+    }
+
+    public get threshold(): number {
```
Comment (JoshuaKGoldberg) [inline src/rules/maxNestingDepthRule.ts]: > public

Nit: can this be `private` please? Better to indicate that `threshold` isn't used anywhere else.

File: src/rules/maxNestingDepthRule.ts [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+}
+
+function walk(ctx: Lint.WalkContext<{ threshold: number }>): void {
+    const {
+        options: { threshold },
+    } = ctx;
+    const nestingDepthStack: number[] = [];
+    const functionNameStack: string[] = [];
+    const cbNode = (node: ts.Node): void => {
+        if (isFunctionScopeBoundary(node)) {
```
Comment (JoshuaKGoldberg) [inline src/rules/maxNestingDepthRule.ts]: > `isFunctionScopeBoundary`

I believe you can use `isFunctionWithBody` from `tsutils` so that you don't need the `as` cast a few lines down.

File: src/rules/maxNestingDepthRule.ts [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+        },
+        optionExamples: [true, [true, 5]],
+        type: "maintainability",
+        typescriptOnly: false,
+    };
+    /* tslint:enable:object-literal-sort-keys */
+
+    public static FAILURE_STRING(expected: number, actual: number, name?: string): string {
+        return (
+            `The function${name === "" ? "" : ` ${name}`} has a nesting depth of at least ` +
```
[the remark this answers is not in the document]
Comment (DamienGarrido) [reply src/rules/maxNestingDepthRule.ts]: I used the wording "_has a nesting depth of at least N_" since I stop parsing the function as soon as the maximum nesting depth is reached.
So the function nesting depth is _at least_ equal to the maximum nesting depth, but it might be greater.

File: tslint.json (hunk 1 above)
Comment (DamienGarrido) [reply tslint.json]: Yes, Prettier has formatted the file during the precommit.
I only added the following line:
```typescript
"max-nesting-depth": [false],
```

Comment (JoshuaKGoldberg) [reply tslint.json]: Heh. I thought I'd gotten all of them in #4270, looks like I missed this. Thanks!

File: src/rules/maxNestingDepthRule.ts [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+            A starting value of 0 is assigned, then this value is
+            incremented each time a new block is entered, and decremented
+            each time a block is left.
+            The following statements contribute to nesting depth:
+            * \`if\` statements
+            * \`switch\` statements
+            * \`for\`, \`for in\` and \`for of\` loops
+            * \`while\` and \`do while\` loops
+            * \`try\` \`catch\` and \`finally\` blocks`,
+        rationale: Lint.Utils.dedent`
```
[the remark this answers is not in the document]
Comment (DamienGarrido) [reply src/rules/maxNestingDepthRule.ts]: I added a paragraph in both documentations. Does it fit ?

File: src/rules/maxNestingDepthRule.ts [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+            Maximum nesting depth is a code metric which indicates the level of imbrication in a
```
[the remark this answers is not in the document]
Comment (DamienGarrido) [reply src/rules/maxNestingDepthRule.ts]: :smiley: Sorry, _imbrication_ is commonly used in my mother tongue. I have slightly rewrote the documentation. Tell me what you think.

File: src/rules/maxNestingDepthRule.ts (hunk 2 above)
Comment (DamienGarrido) [reply src/rules/maxNestingDepthRule.ts]: Of course !

File: src/rules/maxNestingDepthRule.ts (hunk 3 above)
Comment (DamienGarrido) [reply src/rules/maxNestingDepthRule.ts]: `isFunctionWithBody` is indeed more appropriate since the needed tests are narrowed, but as it is also returning a boolean I still need the `as` cast.

File: test/rules/max-nesting-depth/invalidThreshold/test.ts.lint [hunk 7]
```diff
@@ -0,0 +1,21 @@
+// Check that rule is not applied for invalid config values by checking a
+// function with complexity higher than the dafault threshold.
```
[the remark this answers is not in the document]
Comment (DamienGarrido) [reply test/rules/max-nesting-depth/invalidThreshold/test.ts.lint]: Fixed, as well in the cyclomatic-complexity metric.

File: src/rules/maxNestingDepthRule.ts (hunk 3 above)
Comment (DamienGarrido) [reply src/rules/maxNestingDepthRule.ts]: Oh! I missed the declaration, you were right!

File: test/rules/max-nesting-depth/specifiedThreshold/test.ts.lint [hunk 8]
```diff
@@ -0,0 +1,270 @@
...
+        const condition = false ? "true" : false ? "true" : "false";
+    }
+}
+
+// Both invalid.
+function outer4() {
+    const condition = false ? "true" : false ? "true" : false ? "true" : "false";
+                                                        ~~~~~~~~~~~~~~~~~~~~~~~~  [The function outer4 has a nesting depth of at least 3 which is higher than the threshold of 2]
+   function inner4() {
+        const condition = false ? "true" : false ? "true" : false ? "true" : "false";
```
Comment (giladgray) [inline test/rules/max-nesting-depth/specifiedThreshold/test.ts.lint]: surprised that the error here is at the same depth as line 264, given that this function is nested one level deeper. why is that?

File: src/rules/maxNestingDepthRule.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (JoshuaKGoldberg) [reply src/rules/maxNestingDepthRule.ts]: Looks great, thanks!

File: src/rules/maxNestingDepthRule.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (JoshuaKGoldberg) [reply src/rules/maxNestingDepthRule.ts]: Great, thanks!

File: src/rules/maxNestingDepthRule.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (JoshuaKGoldberg) [reply src/rules/maxNestingDepthRule.ts]: On second thought - I think it could be a little frustrating for users to not know how far away their function's nesting depth is. Not explicitly stating the nesting depth for functions also means it'll be hard for folks to experimentally figure out what syntax types increase it.

Thoughts @giladgray?

[the remark this answers is not in the document]
Comment (DamienGarrido) [reply src/rules/maxNestingDepthRule.ts]: I chose to stop parsing the AST as soon as possible after reading these hints: https://palantir.github.io/tslint/develop/custom-rules/performance-tips.html
But I could change the implementation to display a message for each statement deeper than the maximum allowed depth (as ESLint does).

File: test/rules/max-nesting-depth/specifiedThreshold/test.ts.lint (hunk 8 above)
Comment (DamienGarrido) [reply test/rules/max-nesting-depth/specifiedThreshold/test.ts.lint]: It is so because, as ESLint does, I reset the depth counter each time a new function is entered.

Comment (DamienGarrido) [reply test/rules/max-nesting-depth/specifiedThreshold/test.ts.lint]: Does it make sense ?

File: src/rules/maxNestingDepthRule.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (DamienGarrido) [reply src/rules/maxNestingDepthRule.ts]: Happy new year !
Any update @JoshuaKGoldberg?
Any thoughts @giladgray?

[the remark this answers is not in the document]
Comment (JoshuaKGoldberg) [reply src/rules/maxNestingDepthRule.ts]: Happy new year! 🎉

Waiting on the implementation to mention the real nesting depth number.

**Comment by adidahiya** [comment]
looks like there is one unresolved CR feedback item and there are some merge conflicts

File: src/rules/maxNestingDepthRule.ts [hunk 9]
```diff
@@ -0,0 +1,183 @@
...
+            default:
+                return false;
+        }
+    }
+
+    /**
+     * Walks the AST
+     * @param ctx Walk context
+     */
+    public static walk(ctx: Lint.WalkContext<{ threshold: number }>): void {
```
Comment (JoshuaKGoldberg) [inline src/rules/maxNestingDepthRule.ts]: It's unusual for rules to have a `public static walk...` instead of a standalone `const walk = ...`. There's no real need to put this as a static method on the class. We generally prefer standalone functions to lower indentation amounts and decouple ourselves from classes.

File: src/rules/maxNestingDepthRule.ts [hunk 10]
```diff
@@ -0,0 +1,183 @@
...
+        const {
+            options: { threshold },
+        } = ctx;
+        const nestingDepthStack: number[] = [];
+        const functionNameStack: string[] = [];
+        const cbNode: (node: ts.Node) => void = (node: ts.Node): void => {
```
Comment (JoshuaKGoldberg) [inline src/rules/maxNestingDepthRule.ts]: Nit: the `: (node: ts.Node) => void` typedef seems unnecessary and makes it a little harder to read through this function. Let's remove it?

```suggestion
        const cbNode = (node: ts.Node): void => {
```

File: src/rules/maxNestingDepthRule.ts [hunk 11]
```diff
@@ -0,0 +1,183 @@
...
+        rationale: Lint.Utils.dedent`
+            Maximum nesting depth is a code metric which indicates how deep control statements are
+            nested in a function. This metric quickly reveals the functions whose code is made
+            difficult to read by an excessive nesting of the control statements.
+
+            It's better to have smaller, single-purpose functions with self-documenting names.`,
+        optionsDescription: Lint.Utils.dedent`
+            An optional upper limit for maximum nesting depth can be specified. If no limit option
+            is provided a default value of ${Rule.DEFAULT_THRESHOLD} will be used.`,
+        options: {
```
Comment (JoshuaKGoldberg) [inline src/rules/maxNestingDepthRule.ts]: TSLint's been moving towards having options as objects with members, rather than single primitives. They're a lot easier to add more options to later on and a bit more clear in `tslint.json`s. You'll want to switch this to `type: "object"`; see `memberOrderingRule.ts` and `noShawdowedVariableRule.ts` as examples.

File: src/configs/recommended.ts [hunk 12]
```diff
@@ -63,6 +63,7 @@ export const rules = {
     "max-line-length": {
         options: [120],
     },
+    "max-nesting-depth": true,
```
Comment (JoshuaKGoldberg) [inline src/configs/recommended.ts]: Adding this to `recommended.ts` would be a breaking change. I'll remove it. We can look into adding it back in for the next major release, though I don't know that the importance of enabling this rule trumps the annoyance some users would have seeing errors from it on existing code.

**Comment by JoshuaKGoldberg** [comment]
👋 @DamienGarrido just a heads up - I've resolved all PR feedback except for asking the rule to log the full function stacks, rather than an "at least" number. I don't quite have the time to continue through that. Closing the PR for housekeeping, but if you do still want to have this merged (see also #4534), please do feel free to push more!
